torchtext/experimental/vocab.py

```diff
@@ -0,0 +1,145 @@
+import logging
+from typing import Dict, List
+
+import torch
+import torch.nn as nn
+
+logger = logging.getLogger(__name__)
+
+
+class Vocab(nn.Module):
+    r"""Creates a vocab object which maps tokens to indices.
+
+    Note that the ordering in which key value pairs were inserted in the `ordered_dict` will be respected when building the vocab.
+    Therefore if sorting by token frequency is important to the user, the `ordered_dict` should be created in a way to reflect this.
+
+    Arguments:
+        ordered_dict (collections.OrderedDict): object holding the frequencies of each token found in the data.
+        min_freq: The minimum frequency needed to include a token in the vocabulary.
+            Values less than 1 will be set to 1. Default: 1.
+        specials: The tuple of special tokens (e.g., padding or eos) that will be prepended/postpended to the vocabulary.
+            based on the `specials_first` flag. The ordering of the tuple will be preserved. Default: ('<unk>', '<pad>')
+        specials_first: Whether to add special tokens into the vocabulary at first. If it is False,
+            they are added into the vocabulary at last. Default: True.
+
+    Raises:
+        ValueError: if a default `unk_token` isn't provided.
+    """
+
+    def __init__(self, ordered_dict, min_freq=1, unk_token='<unk>', specials=('<unk>', '<pad>'), specials_first=True):
+        super(Vocab, self).__init__()
+
+        if not unk_token:
+            raise ValueError("A default unk token wasn't provided.")
+
+        if unk_token not in specials:
+            raise ValueError("The unk token wasn't found in the `specials` tuple.")
+
+        tokens = []
+        for token, freq in ordered_dict.items():
+            if freq >= min_freq:
+                if token in specials:
+                    raise ValueError("A `specials` token {} was found inside of `ordered_dict`."
+                                     "Please ensure that the `ordered_dict` doesn't contain any special tokens.".format(token))
+                tokens.append(token)
+
+        # assume special tokens dont appear in ordered_dict
+        if specials_first:
+            tokens = list(specials) + tokens
+        else:
+            tokens += list(specials)
+
+        self.vocab = torch.classes.torchtext.Vocab(tokens, unk_token)
+
+    @torch.jit.export
+    def __len__(self) -> int:
+        r"""Returns:
+            length (int): the length of the vocab
+        """
+        return len(self.vocab)
+
+    @torch.jit.export
+    def __getitem__(self, token: str) -> int:
+        r"""
+        Args:
+            token (str): the token used to lookup the corresponding index.
+
+        Returns:
+            index (int): the index corresponding to the associated token.
+        """
+        return self.vocab[token]
+
+    @torch.jit.export
+    def insert_token(self, token: str, index: int) -> None:
```

Comment:
Do we want to maintain `insert_token` func in vocab class? Any usage case to support it?

Reply:
Yes I think Steven specifically requested this during our design meeting. I.e. they wanted to be able to insert special tokens into specific indices after constructing a `Vocab`.